File: step_13_binary_trees/13.3_hard/P014_flatten_bt_to_linked_list/Solution.java
/**
 * Problem: Flatten Binary Tree to Linked List (LeetCode 114)
 * Difficulty: MEDIUM | XP: 25
 *
 * Flatten a binary tree in-place into a right-only "linked list"
 * following preorder traversal order. Left pointers must be null.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ------------------------------------------------------------------
    // Shared TreeNode definition
    // ------------------------------------------------------------------
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int v) { val = v; }
        TreeNode(int v, TreeNode l, TreeNode r) { val = v; left = l; right = r; }
    }

    // Helper: build tree from level-order array (null = missing child)
    static TreeNode build(Integer[] vals) {
        if (vals == null || vals.length == 0) return null;
        TreeNode root = new TreeNode(vals[0]);
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int i = 1;
        while (!q.isEmpty() && i < vals.length) {
            TreeNode node = q.poll();
            if (i < vals.length && vals[i] != null) { node.left = new TreeNode(vals[i]); q.offer(node.left); }
            i++;
            if (i < vals.length && vals[i] != null) { node.right = new TreeNode(vals[i]); q.offer(node.right); }
            i++;
        }
        return root;
    }

    // Helper: collect flattened chain into list for display
    static List<Integer> toList(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        for (TreeNode n = root; n != null; n = n.right) res.add(n.val);
        return res;
    }

    // ============================================================
    // APPROACH 1: BRUTE FORCE — Preorder to array, then relink
    // Time: O(n)  |  Space: O(n)
    // ============================================================
    static class BruteForce {
        /**
         * Collect all nodes in preorder into a list, then wire each node's
         * right pointer to the next node and null out all left pointers.
         * Simple and readable but uses O(n) extra space for the node list.
         */
        List<TreeNode> nodes = new ArrayList<>();

        void preorder(TreeNode node) {
            if (node == null) return;
            nodes.add(node);
            preorder(node.left);
            preorder(node.right);
        }

        void flatten(TreeNode root) {
            preorder(root);
            for (int i = 0; i < nodes.size() - 1; i++) {
                nodes.get(i).left = null;
                nodes.get(i).right = nodes.get(i + 1);
            }
            if (!nodes.isEmpty()) {
                nodes.get(nodes.size() - 1).left = null;
                nodes.get(nodes.size() - 1).right = null;
            }
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL — Recursive reverse-postorder
    // Time: O(n)  |  Space: O(h) call stack
    // ============================================================
    static class Optimal {
        /**
         * Process nodes in reverse preorder: right → left → root.
         * A 'prev' pointer is maintained. After processing both subtrees,
         * set node.right = prev and node.left = null, then update prev = node.
         * When we reach the root, the entire tree is already flattened on the right.
         */
        TreeNode prev = null;

        void flatten(TreeNode root) {
            if (root == null) return;
            flatten(root.right);
            flatten(root.left);
            root.right = prev;
            root.left = null;
            prev = root;
        }
    }

    // ============================================================
    // APPROACH 3: BEST — Morris-style iterative O(1) space
    // Time: O(n)  |  Space: O(1)
    // ============================================================
    static class Best {
        /**
         * At each node that has a left child:
         *   1. Find the rightmost node of the left subtree (the preorder predecessor).
         *   2. Thread the current right subtree onto that predecessor's right.
         *   3. Move the left subtree to the right, null out left.
         * Advance to curr.right. No stack or recursion needed.
         * This mirrors Morris traversal and is truly O(1) extra space.
         */
        void flatten(TreeNode root) {
            TreeNode curr = root;
            while (curr != null) {
                if (curr.left != null) {
                    // Find rightmost of left subtree
                    TreeNode pred = curr.left;
                    while (pred.right != null) pred = pred.right;
                    // Thread right subtree
                    pred.right = curr.right;
                    // Shift left to right
                    curr.right = curr.left;
                    curr.left = null;
                }
                curr = curr.right;
            }
        }
    }

    // ============================================================
    // MAIN
    // ============================================================
    public static void main(String[] args) {
        System.out.println("=== Flatten Binary Tree to Linked List ===\n");
        // Tree: 1->2->5, 2->3->4, 5->6
        // Expected: [1, 2, 3, 4, 5, 6]
        Integer[] vals = {1, 2, 5, 3, 4, null, 6};

        TreeNode r1 = build(vals);
        new BruteForce().flatten(r1);
        System.out.println("Brute:         " + toList(r1));

        TreeNode r2 = build(vals);
        new Optimal().flatten(r2);
        System.out.println("Optimal:       " + toList(r2));

        TreeNode r3 = build(vals);
        new Best().flatten(r3);
        System.out.println("Best (Morris): " + toList(r3));
    }
}

File: step_13_binary_trees/13.3_hard/P014_flatten_bt_to_linked_list/solution.py
"""
Problem: Flatten Binary Tree to Linked List (LeetCode 114)
Difficulty: MEDIUM | XP: 25

Flatten a binary tree into a linked list in-place using preorder traversal order.
The "linked list" uses the right pointer; left pointers must be null.
"""
from typing import List, Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values: List[Optional[int]]) -> Optional[TreeNode]:
    """Helper: build tree from level-order list."""
    if not values:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root: Optional[TreeNode]) -> List[int]:
    """Helper: read flattened right-only chain."""
    result = []
    node = root
    while node:
        result.append(node.val)
        node = node.right
    return result


# ============================================================
# APPROACH 1: BRUTE FORCE
# Collect all nodes in preorder into a list, then relink.
# Time: O(n)  |  Space: O(n)
# ============================================================
def flatten_brute(root: Optional[TreeNode]) -> None:
    """
    Intuition: Do a standard preorder traversal to capture the visit
    sequence, then wire each node's right to the next node and left to None.
    Simple but uses O(n) extra memory.
    """
    nodes: List[TreeNode] = []

    def preorder(node: Optional[TreeNode]) -> None:
        if node is None:
            return
        nodes.append(node)
        preorder(node.left)
        preorder(node.right)

    preorder(root)

    for i in range(len(nodes) - 1):
        nodes[i].left = None
        nodes[i].right = nodes[i + 1]
    if nodes:
        nodes[-1].left = None
        nodes[-1].right = None


# ============================================================
# APPROACH 2: OPTIMAL — Recursive reverse-postorder (right → left → root)
# Time: O(n)  |  Space: O(h) call stack
# ============================================================
def flatten_optimal(root: Optional[TreeNode]) -> None:
    """
    Intuition: Process nodes in reverse preorder (right subtree → left subtree
    → root). Keep a 'prev' pointer. When we visit the root last, we set
    root.right = prev and root.left = None. This stitches the list without
    any extra array.

    Trick: visiting in reverse order (right first, then left) means when we
    process the root, 'prev' already points to the head of the already-flattened
    right subtree — perfect for right-linking.
    """
    prev: List[Optional[TreeNode]] = [None]  # use list to allow mutation in closure

    def dfs(node: Optional[TreeNode]) -> None:
        if node is None:
            return
        dfs(node.right)
        dfs(node.left)
        node.right = prev[0]
        node.left = None
        prev[0] = node

    dfs(root)


# ============================================================
# APPROACH 3: BEST — Morris-style iterative O(1) space
# Time: O(n)  |  Space: O(1)
# ============================================================
def flatten_best(root: Optional[TreeNode]) -> None:
    """
    Intuition: At each node with a left child, find the rightmost node of the
    left subtree (the predecessor in preorder). Thread the current node's right
    subtree onto that predecessor's right pointer, then move the left subtree
    to the right and null out the left pointer. Advance to root.right. Repeat.

    This is the same idea as Morris traversal — no stack, no recursion.
    """
    curr = root
    while curr:
        if curr.left:
            # Find rightmost node of left subtree
            predecessor = curr.left
            while predecessor.right:
                predecessor = predecessor.right
            # Thread right subtree onto predecessor
            predecessor.right = curr.right
            # Move left subtree to right
            curr.right = curr.left
            curr.left = None
        curr = curr.right


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Flatten Binary Tree to Linked List ===\n")

    # Tree:      1
    #           / \
    #          2   5
    #         / \   \
    #        3   4   6
    # Expected preorder: [1, 2, 3, 4, 5, 6]

    for name, fn in [("Brute", flatten_brute),
                     ("Optimal", flatten_optimal),
                     ("Best (Morris)", flatten_best)]:
        root = build_tree([1, 2, 5, 3, 4, None, 6])
        fn(root)
        print(f"{name:15s}: {tree_to_list(root)}")
